HSEARCH-798 Use logging categories

marko-bekhta opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-798

This commit only splits the existing `Log` interfaces and moves messages/exceptions to more "specific" interfaces that represent logging categories.
While at it I've adjusted the usage of logging interfaces in a way that they are not used outside the module (I'll take a look at jqassistant to see if any rules can be adjusted there) as I've noticed that some were "shared" even if just for the tests.

Also, there were some string messages in the logging interfaces. Those are now moved to new/existing `*Hints`  message bundles

It also adds a new annotation:

```java
@CategorizedLogger(
		category = FormattingLog.CATEGORY_NAME,
		description = "Anything related to parsing/formatting."
)
```

which may change in near future 😄. The plan is to have all the interfaces-categories be marked with this interface and then collect the info from it to build a documentation appendix containing available logging categories and what they are for. Not all categories are used for logging (some are there only with exceptions), hence the plan is to check:
- is there any `void blabla(..)` methods 
- or whether the interface-category extends the `BasicLogger` (and potentially uses `SOME_LOGGER.infof` kind of methods) 

And if so, consider that category as such that can log something. 
Then, since categories are shared between modules each module may contribute more to the category description.

To address the "hmmmmmmmmmmmmmmmm what's the next id to use?!!" I've added `nextLoggerIdForConvenience` in the interfaces combining categories. It has the next "free ID"; let's see how it goes in practice...

Next, I'll work on "something" that generates that docs appendix. 

<!--
Please read and do not remove the following lines:
-->
----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-search/blob/main/CONTRIBUTING.md#legal).

----------------------

marko-bekhta commented:
I've looked at the reports that the logging lib produces: 
![image](https://github.com/user-attachments/assets/e9678a51-36a9-421e-ae05-8cc528a8929a)

and it doesn't match what we'd wanted to add to the reference. Hence, I ended up with an annotation processor that generates a YAML into `meta-inf/hibernate-search/logging-categories.yaml` file like:
```yaml
# List of logging categories that this module utilizes.
org.hibernate.search.report:
  module: hibernate-search-engine
  categories:
  - name: org.hibernate.search.common.failures
    description:
    - |
      Logs unexpected problems that occur during background Hibernate Search processes.
  - name: org.hibernate.search.configuration
    description:
    - |
      Logs warnings about unused configuration properties and tracking of such properties.
      +
      Debug logs may also include information on configuration providers or other configuration-related issues.
  - name: org.hibernate.search.executor
    description:
    - |
      Logs the information on various executor operations.
      Executor in this case in particular can be the one used to batch the index work.
```

Then in the documentation module, there's a new exec plugin execution that collects all these yamls and generates an adoc file added as an appendix: 
![image](https://github.com/user-attachments/assets/c5297eae-76aa-40be-bb9b-41d70bcf02f6)

To try this out .... 🙈 

```bash
mvn clean install -am -pl documentation -DskipTests
```

Logger descriptions would need some more attention, but I thought.. I'll see to that once the list of logging categories is "finalized".

NOTE: the annotation processor generates that yaml into meta inf so that we wouldn't have troubles with Develocity, like if we'd generate a file somewhere in the root (on compile goal...)

@yrodiere, I'll keep this PR as is for now so you can have a look at it when you have time 😃.

yrodiere reviewed (commented):
I (obviously) didn't review everything, but here are a few comments... I trust you'll pick what makes sense, ignore the rest, and will know to extrapolate to other modules :)

---

Regarding the YAML, I think `description` shouldn't be a list just text, no? I.e.:

> ```diff
>  # List of logging categories that this module utilizes.
>  org.hibernate.search.report:
>    module: hibernate-search-engine
>    categories:
>    - name: org.hibernate.search.common.failures
> -     description:
> -    - |
> +     description: |
>        Logs unexpected problems that occur during background Hibernate Search processes.
> ```

Depending how you transform that into asciidoc, this could also get rid of the extra bullet points under "Description" in the documentation...

marko-bekhta commented:
Thanks for taking a look! Will go through your review shortly 😃 🥳

marko-bekhta commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/ElasticsearchLog.java:
```diff
@@ -0,0 +1,46 @@
+		@ValidIdRange(min = MessageConstants.BACKEND_ES_ID_RANGE_MIN, max = MessageConstants.BACKEND_ES_ID_RANGE_MAX),
+})
+public interface ElasticsearchLog
+		extends AnalyzerLog, CommonFailureLog, ConfigurationLog, ElasticsearchClientLog, ElasticsearchRequestLog,
+		ElasticsearchSpecificLog, IndexingLog,
+		MappingLog, QueryLog, SchemaExportLog, VersionLog {
```
I'll probably better reply to some of these before finishing the renaming to `INSTANCE` 😄 

> If you really want to go that far, wouldn't it make sense to have all of these as nested interfaces under ElasticsearchLog, at least, so that it's easier to maintain -- especially when it comes to the ID range?

Yeah that was my initial thought as well to keep them all as inner interfaces but:

>Cyclic inheritance involving 'org. hibernate. search. engine. logging. impl. EngineLog'

If we want the main interface to extend these subs... and I think we do want it to extend them as that way, we rely on the AP to verify that there's no message ID overlap... 

Note that the category collector will only include the categories where the interface has the `@LogMessage(level = ` so if we have interfaces with just exceptions, those won't be listed as logging categories.

marko-bekhta commented on backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/logging/impl/ElasticsearchRequestLog.java:
```diff
@@ -0,0 +1,76 @@
+	 * <p>
+	 * To enable the logger, the category needs to be enabled at TRACE level.
+	 */
+	String CATEGORY_NAME = "org.hibernate.search.backend.elasticsearch.request";
```
I was going back and forth with this one 🙈 we also already had a `org.hibernate.search.backend.lucene.infostream` that had backend in it ... so I was like.. add a backend to the Elasticsearch category ... drop `.backend` from the infostream one? It felt inconsistent to keep it as is 🙈 

> so that people can just enable debug of the client and automatically get request logging?

I'll check if it'll work like that... I thought it only works for "exact match" but even if it doesn't the suggestion still makes sense to move request part after client 👍🏻

marko-bekhta commented:
> Regarding the YAML, I think description shouldn't be a list just text, no? I.e.:

The initial idea was that since a category can be shared between multiple modules each module can produce multiple descriptions for the same category. So I thought to keep it "flexible" and allow multiple descriptions from the same module (even though it doesn't make that much sense, but it was letting make it simple on the report rendering side). But now since there are "module suffixes" if the category is shared this is no longer needed. So, the report contains the description as a single text block, and the processor should fail the build if the same category is used multiple times within the same module. 

I think I've addressed most of your suggestions (except the one on the interface grouping in a single parent interface since that doesn't pass the build). I'll merge this PR as there's a few files changed in it 😄 and rebasing each time requires resolving conflicts 🙈. If there are any other ideas or suggestions, I'll address them in follow-ups.
Thanks one more time for taking a look at the PR!
